File: 17_letter_combinations_of_a_phone_number/letterCombinations.py
class Solution(object):
    def letterCombinations(self, digits):
        num_dic = {'1':'*', '2': 'abc', '3': 'def',
                    '4': 'ghi', '5': 'jkl', '6': 'mno',
                    '7': 'pqrs', '8':'tuv', '9': 'wxyz',
                    '0': ' '}
        digits = list(digits)
        com_len = len(digits)
        result = []

        def dfs(s, start):
            if len(s) == com_len: 
                result.append(s)
                return 

            for word in num_dic[digits[start]]:
                dfs(s + word, start + 1)


        dfs('', 0)
        return result

sol = Solution()
print sol.letterCombinations('23')

File: 17_letter_combinations_of_a_phone_number/letterCombinations.cpp
class Solution {
public:
    vector<string> letterCombinations(string digits) {
        vector<string> res;
        // corner case
        if(digits.size() == 0) { return res; }
        numMap = {"0", "1", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
        string s = "";
        letterCombinationsHelper(res, digits, s, 0);
        return res;
    }
private:
    vector<string> numMap;

    void letterCombinationsHelper(vector<string>& res, string& digits, string& s, int step) {
        if(step >= digits.size()) {
            res.push_back(s);
            return;
        }

        string chars = numMap[digits[step] - '0'];

        for(char c : chars) {
            s.push_back(c);
            letterCombinationsHelper(res, digits, s, step + 1);
            s.pop_back();
        }
    }
};